public/dorequest.php

```diff
@@ -575,7 +578,38 @@ function LogDeprecatedUserAgent(string $requestType, ClientSupportLevel $clientS
 
     case "patch":
         $flag = (int) request()->input('f', 0);
-        $response = GetPatchData($gameID, $user, $flag);
+        $gameHashMd5 = request()->input('m');
+
+        $clientSupportLevel = (new GetClientSupportLevelAction())->execute(request()->header('User-Agent'));
+
+        // TODO middleware?
+        if ($clientSupportLevel === ClientSupportLevel::Blocked) {
+            return DoRequestError('This client is not supported', 403, 'unsupported_client');
+        }
+
+        try {
+            $gameHash = $gameHashMd5 ? GameHash::whereMd5($gameHashMd5)->first() : null;
+            $game = $gameHashMd5 ? null : Game::find($gameID);
+
+            $response = (new BuildClientPatchDataAction())->execute(
+                gameHash: $gameHash,
+                game: $game,
+                user: $user,
+                flag: AchievementFlag::tryFrom($flag),
+            );
+
+            // Based on the user's current client support level, we may want to attach
+            // some metadata into the patch response. We'll do that as part of a separate
+            // action to keep the original data construction pure.
+            $response = (new InjectPatchClientSupportLevelDataAction())->execute(
+                $response,
+                $clientSupportLevel,
+                $gameHash,
+                $game,
+            );
```

Comment:
This is how I integrated the new stuff from #2862.

First, execute `BuildClientPatchDataAction`, which always should build the correct patch package for the game regardless of the user's client.

Next, call `InjectPatchClientSupportLevelDataAction` on the returned patch data. This is almost like a side-effect, injecting whatever warning achievements and warning metadata we may want to send to a client.

Reply:
I wonder if it would make sense to send the warning back as a new field if the input is a hash. There aren't any clients sending hashes yet, so we could structure the warning more appropriately instead of shoehorning in into something that existing clients would understand.

Reply:
Definitely not opposed to this. I don't want to be presumptuous - what would you prefer in the response in this circumstance?

Reply:
`rc_client` doesn't currently have any way to plumb the warning through, so we could design whatever is appropriate. 

The closest mechanism currently available would be to raise a [SERVER_ERROR](https://github.com/RetroAchievements/rcheevos/wiki/rc_client_set_event_handler#rc_client_event_server_error) event, which would end up displaying something like "patch failed: Outdated Emulator (please update)" in RetroArch. I'm not sure how other clients handle the event.

The more I think about it, the more I like using an achievement to report the warning. In addition to ensuring a high level of compatibility across clients (they'd be much more likely to handle the `ACHIEVEMENT_TRIGGERED` event than the `SERVER_ERROR` event), it keeps the warning around in the achievement list so users can be reminded while they're in the game.

Reply:
> it keeps the warning around in the achievement list so users can be reminded while they're in the game.

This is a good point. It sounds like it may be best to keep it as-is for now then.